src/pip/_internal/cli/main_parser.py

```diff
@@ -92,8 +103,6 @@ def parse_command(args):
 
         raise CommandError(' - '.join(msg))
 
-    # all the args without the subcommand
-    cmd_args = args[:]
-    cmd_args.remove(cmd_name)
-
-    return cmd_name, cmd_args
+    if set(['--user', '--target']).issubset(set(args_else)):
```

Comment:
I would handle this in `InstallCommand` instead, since it is the only context where `--user` and `--target` are mutually exclusive.

It is much easier to handle it in `run()` as well; `optparse` documentation [conveniently provides an example](https://docs.python.org/3/library/optparse.html#how-optparse-handles-errors); you can simply write something like

```python
def run(self, options, args):
    if options.use_user_site and options.target_dir is not None:
        self.parser.error('--user and --target cannot be used together.')
    # ...
```

Reply:
@uranusjr But why do you wanna run all this code unnecessarily when you already know it's not a valid command. Wasn't that the idea to error out early if both the arguments `--user` and `--target` are used together?

I mean yes handling it in run will be easier. But I would avoid executing the code to populate options and then checking it when I already know these both args aren't allowed together.

Reply:
@uranusjr My bad. I didn't realise `--target` was being used with install command only. But don't you think error out early would be better than executing all the code to populate the options, import install module etc. and then error it out. I mean we already know these both args are not allowed together.

Reply:
> I mean we already know these both args are not allowed together.

Indeed, but you’re already missing `-t` which is an alias of `--target`. And what happens if we add more aliases to either argument? Populating the whole options is actually a good thing IMO since it provides both consistency and safety.

Reply:
Oh yeah that makes sense, Looks like I didn't consider all the cases. Well, will make the changes. Thanks for the review.

Reply:
As we are on the topic. Can you suggest a better error message to show?

Reply:
That I have absolutely no suggestions. I think the current message is good enough, but I’m terrible at this 😆